File: katrinconsole/katrinconsole-web/services/sanity.ts
import { ApiBase, } from './api-base'

export interface AdeiGroup {
    db_server: string
    db_name: string
    db_group: string
    name: string
}

export interface AdeiChannel {
    db_server: string
    db_name: string
    db_group: string
    name: string
    chid: string
    uid: string
    axis: string
}

export interface AdeiDuplicate extends AdeiChannel {
    url: string
}

interface AdeiDuplicateInstance {
    total: number
    channel_uid: string
    duplicates: AdeiDuplicate[]
}

export interface AdeiDuplicates {
    total: number
    unique_channels: number
    duplicates: AdeiDuplicateInstance[]
}

export class SanityService extends ApiBase {
    async groups(): Promise<AdeiGroup[] | undefined> {
        const groups = await this.axiosInstance.get<AdeiGroup[]>('/api/sanity/groups/').catch(() => null)
        return groups?.data
    }

    async channels(): Promise<AdeiChannel[] | undefined> {
        const channels = await this.axiosInstance.get<AdeiChannel[]>('/api/sanity/channels/').catch(() => null)
        return channels?.data
    }

    async duplicates(): Promise<AdeiDuplicates | undefined> {
        const dups = await this.axiosInstance.get<AdeiDuplicates>('/api/sanity/duplicates/').catch(() => null)
        return dups?.data
    }
}

File: katrinconsole/katrinconsole/core/views/sanity.py
import pandas as pd
from django.forms import model_to_dict
from katrinconsole.core.models import AdeiChannel, AdeiGroup
from katrinconsole.core.serializers import (AdeiChannelsSerializer,
                                            AdeiDuplicatesSerializer,
                                            AdeiGroupsSerializer)
from katrinconsole.lib.adei import get_adei_item_url
from rest_framework import generics
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response


class GroupsView(generics.ListAPIView):
    queryset = AdeiGroup.objects.all()
    serializer_class = AdeiGroupsSerializer


class ChannelsView(generics.ListAPIView):
    queryset = AdeiChannel.objects.all()
    serializer_class = AdeiChannelsSerializer


@api_view(['GET'],)
@permission_classes([IsAuthenticated])
def duplicates(req):
    items = list(AdeiChannel.objects.all())
    df = pd.DataFrame([model_to_dict(item) for item in items])
    dup = df[df['uid'].duplicated(keep=False)].groupby('uid')
    total_dup = 0
    unique_dup = 0
    duplicates = []
    for key, items in dup:
        if key:
            length = len(items)
            total_dup += length
            unique_dup += 1
            channel = {
                'total': length,
                'channel_uid': key,
                'duplicates': [{
                    'url': get_adei_item_url(row["db_server"], row["db_name"], row["db_group"]),
                    'uid': row.get('uid', None),
                    'chid': row['chid'],
                    'db_server': row["db_server"],
                    'db_name': row["db_name"],
                    'db_group': row["db_group"],
                } for _, row in items.iterrows()]
            }
        duplicates.append(channel)

    dups = {
        'total': total_dup,
        'unique_channels': unique_dup,
        'duplicates': duplicates
    }
    serializer = AdeiDuplicatesSerializer(dups,)
    return Response(data=serializer.data,)
